20210317 fix npm run coverage

shivamag00 opened the pull request:
Merging this will close #1208 (raised by @jrobbins) 

`npm run converge` now exits without failures & errors 

Many problems occurred when `npm run coverage` was executed:-
- `npm run coverage`  runs `coverage.py` using the local installation of python ( as evident from package.json). Hence, before running this command, one must ensure that coverage is present inside the local installation of python. As suggested by @jrobbins and implemented by @whoanuragverma in #1203 , this can be done using `npm run dev-deps` if coverage is added to  `requirements.dev.txt`
- Even if coverage was present in the local installation of Python, the following errors were raised during the execution of `npm run coverage`.
1. Import Error
```
======================================================================
ERROR: api.cues_api_test (unittest.loader.ModuleImportFailure)
----------------------------------------------------------------------
ImportError: Failed to import test module: api.cues_api_test
Traceback (most recent call last):
  File "/usr/lib/python2.7/unittest/loader.py", line 254, in _find_tests
    module = self._get_module_from_name(name)
  File "/usr/lib/python2.7/unittest/loader.py", line 232, in _get_module_from_name
    __import__(name)
  File "/mnt/c/Users/Shivam/ubuntu/GSOC/chromium-dashboard/tests/api/cues_api_test.py", line 25, in <module>
    from api import cues_api
ImportError: cannot import name cues_api


======================================================================
ERROR: api.stars_api_test (unittest.loader.ModuleImportFailure)
----------------------------------------------------------------------
ImportError: Failed to import test module: api.stars_api_test
Traceback (most recent call last):
  File "/usr/lib/python2.7/unittest/loader.py", line 254, in _find_tests
    module = self._get_module_from_name(name)
  File "/usr/lib/python2.7/unittest/loader.py", line 232, in _get_module_from_name
    __import__(name)
  File "/mnt/c/Users/Shivam/ubuntu/GSOC/chromium-dashboard/tests/api/stars_api_test.py", line 25, in <module>
    from api import register
ImportError: cannot import name register
```

These errors were caused as there was a name conflict between `api` and `tests/api` modules for python files inside `tests` directory. Hence, I rename the `api` to `apio` and kept `tests/api` intact.


2. Assertion Error
```
FAIL: test_get_client__unit_tests (cloud_tasks_helpers_test.CloudTasksHelpersTest)
During unit testing, we are using a fake object.
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/mnt/c/Users/Shivam/ubuntu/GSOC/chromium-dashboard/tests/cloud_tasks_helpers_test.py", line 66, in test_get_client__unit_tests
    type(actual))
AssertionError: <class 'testing_config.FakeCloudTasksClient'> != <class 'tests.testing_config.FakeCloudTasksClient'>
```

This error was caused as `assert.equal()` returned false while comparing `testing_config.FakeCloudTasksClient` & `tests.testing_config.FakeCloudTasksClient`. A closer inspection of the two will reveal that later has `tests.` which is the outer package name. The former did not have `tests.` This was due to importing`testing_config` as `import testing_config`. Importing `testing_config`  using `from tests import testing_config` solved the issue.

whoanuragverma commented on settings.py:
```diff
@@ -64,7 +64,8 @@
 
-SECRET_KEY = os.environ['DJANGO_SECRET']
+# SECRET_KEY = os.environ['DJANGO_SECRET']
+SECRET_KEY = "test secret"
```
Hard Coding secrets in the code itself doesn't make sense. Also this is not a good practise.

shivamag00 replied:
I forgot not to add this change. Thank you @whoanuragverma. I will change it

jrobbins commented:
Thanks for this CL.  Hang on to it for now and you can land a simpler version of it later.

I have been thinking for a while that having a parallel code hierarchy under the "tests/" directory is a bad idea.  Instead, I want to organize the application code into meaningful directories, with each directory being smaller.  Then, I will place the source file X.py and unit tests X_test.py together in the same directory.  That should also eliminate the ambiguity with imports.  I will try to get that done in the next few days.

jrobbins commented:
FYI, I have made some progress on reorganizing unit test files and I am track to complete it in a few days as I said.

jrobbins commented:
I am still working on refactoring code. However, I think that you could redo this PR (or a new one) and achieve the fix much more cleanly now.  Specially, I think that you would only need to edit requirements.dev.txt and package.json. (and also requirements.txt to remove unneeded coverage dev from there).

shivamag00 commented:
@jrobbins, as you suggested, I only had to make the following changes:-

1. added` coverage` dependency to `requirements.dev.txt`
2. modified the script `npm run coverage` to install the dependencies from `requirements.dev.txt`
3. removed `coverage` from `requirements.txt`

Also, as you suggested earlier, after refactoring the files, we ended up with a much simpler version of the PR that was raised earlier.

jrobbins commented on package.json:
```diff
@@ -11,7 +11,7 @@
     "dev-deps": "pip install -r requirements.dev.txt --upgrade",
     "test": "python -m unittest discover -p *_test.py  -b",
-    "coverage": "python -m coverage erase && python -m coverage run -m unittest discover -p *_test.py -s tests -b && python -m coverage html",
+    "coverage": "python -m pip install --no-deps -r requirements.dev.txt && python -m coverage erase && python -m coverage run -m unittest discover -p *_test.py -s tests -b && python -m coverage html",
```
I think you need to remove the `-s tests` because we no longer have a tests directory.

Also, today I landed a change that uses `-p "*_tests.py"` in quotes to avoid shell wildcard expansion if there happens to be a match in the current directory.  So, you just need to have that same change.

shivamag00 replied:
@jrobbins I have made the above-mentioned changes

jrobbins commented:
This will be useful.  Thanks for working through these rounds of review.

AnA025 commented:
> Merging this will close #1208 (raised by @jrobbins) 
> 
> `npm run converge` now exits without failures & errors 
> 
> Many problems occurred when `npm run coverage` was executed:-
> - `npm run coverage`  runs `coverage.py` using the local installation of python ( as evident from package.json). Hence, before running this command, one must ensure that coverage is present inside the local installation of python. As suggested by @jrobbins and implemented by @whoanuragverma in #1203 , this can be done using `npm run dev-deps` if coverage is added to  `requirements.dev.txt`
> - Even if coverage was present in the local installation of Python, the following errors were raised during the execution of `npm run coverage`.
> 1. Import Error
> ```
> ======================================================================
> ERROR: api.cues_api_test (unittest.loader.ModuleImportFailure)
> ----------------------------------------------------------------------
> ImportError: Failed to import test module: api.cues_api_test
> Traceback (most recent call last):
>   File "/usr/lib/python2.7/unittest/loader.py", line 254, in _find_tests
>     module = self._get_module_from_name(name)
>   File "/usr/lib/python2.7/unittest/loader.py", line 232, in _get_module_from_name
>     __import__(name)
>   File "/mnt/c/Users/Shivam/ubuntu/GSOC/chromium-dashboard/tests/api/cues_api_test.py", line 25, in <module>
>     from api import cues_api
> ImportError: cannot import name cues_api
> 
> 
> ======================================================================
> ERROR: api.stars_api_test (unittest.loader.ModuleImportFailure)
> ----------------------------------------------------------------------
> ImportError: Failed to import test module: api.stars_api_test
> Traceback (most recent call last):
>   File "/usr/lib/python2.7/unittest/loader.py", line 254, in _find_tests
>     module = self._get_module_from_name(name)
>   File "/usr/lib/python2.7/unittest/loader.py", line 232, in _get_module_from_name
>     __import__(name)
>   File "/mnt/c/Users/Shivam/ubuntu/GSOC/chromium-dashboard/tests/api/stars_api_test.py", line 25, in <module>
>     from api import register
> ImportError: cannot import name register
> ```
> 
> These errors were caused as there was a name conflict between `api` and `tests/api` modules for python files inside `tests` directory. Hence, I rename the `api` to `apio` and kept `tests/api` intact.
> 
> 
> 2. Assertion Error
> ```
> FAIL: test_get_client__unit_tests (cloud_tasks_helpers_test.CloudTasksHelpersTest)
> During unit testing, we are using a fake object.
> ----------------------------------------------------------------------
> Traceback (most recent call last):
>   File "/mnt/c/Users/Shivam/ubuntu/GSOC/chromium-dashboard/tests/cloud_tasks_helpers_test.py", line 66, in test_get_client__unit_tests
>     type(actual))
> AssertionError: <class 'testing_config.FakeCloudTasksClient'> != <class 'tests.testing_config.FakeCloudTasksClient'>
> ```
> 
> This error was caused as `assert.equal()` returned false while comparing `testing_config.FakeCloudTasksClient` & `tests.testing_config.FakeCloudTasksClient`. A closer inspection of the two will reveal that later has `tests.` which is the outer package name. The former did not have `tests.` This was due to importing`testing_config` as `import testing_config`. Importing `testing_config`  using `from tests import testing_config` solved the issue.
